packages/mux-player/src/utils.js

```diff
@@ -0,0 +1,162 @@
+        template
+      );
+    }
+    fragment.id = id || i;
+    return fragment;
+  });
+
+  // Replace the placeholders with the fragments.
+  childFragments.forEach((fragment, i) => {
+    const df = new DocumentFragment();
+    if (fragment instanceof PersistentFragment) {
+      df.append(...fragment.childNodes);
+    } else {
+      // Could be a Node or any other object.
+      df.append(fragment);
+    }
+    const placeholder = compiler.content.querySelector(`#_${uid}${i + 1}`);
+    placeholder.parentNode.replaceChild(df, placeholder);
+  });
+
+  return new PersistentFragment(compiler.content.childNodes, childFragments);
+}
+
+class PersistentFragment {
+  /**
+   * Create a new PersistentFragment.
+   * A normal DocumentFragment loses its children when appended to the document.
+   * PersistentFragment keeps track of its children and makes them easy to replace.
+   *
+   * @param  {Node[]} childNodes
+   * @param  {PersistentFragment[]} childFragments
+   */
+  constructor(childNodes, childFragments) {
+    this.childFragments = childFragments;
+    this._cachedChildNodes = [...childNodes];
+    this._mark = new Text();
+    this.id = undefined;
+  }
+
+  /**
+   * Return a map with child fragments.
+   * @return {Object.<string, PersistentFragment>}
+   */
+  get fragments() {
+    let map = {};
+    for (let frag of this.childFragments) map[frag.id] = frag;
+    return map;
+  }
+
+  get childNodes() {
+    return [this._mark, ...this._cachedChildNodes];
+  }
+
+  replaceChildren(...nodesOrDOMStrings) {
+    const childNodes = this._cachedChildNodes;
+    this._cachedChildNodes = [...nodesOrDOMStrings];
+
+    childNodes.forEach((n) => n.remove());
+    let next = this._mark.nextSibling;
+    nodesOrDOMStrings.forEach((node) =>
+      this._mark.parentNode.insertBefore(node, next)
+    );
+  }
+}
+
+class RenderableFragment extends PersistentFragment {
+  /**
+   * Create a new RenderableFragment.
+   * The problem RenderableFragment solves is to be able to define inline
+   * render functions in the template literal and then later re-render those
+   * by just passing props.
+   * @param  {Node[]} childNodes
+   * @param  {PersistentFragment[]} childFragments
+   * @param  {(props?: Object) => PersistentFragment} template
+   */
+  constructor(childNodes, childFragments, template) {
+    super(childNodes, childFragments);
+    this.template = template;
+  }
+
+  render(props) {
+    const newFragment = this.template(props) || html``;
+    this.childFragments = newFragment.childFragments;
+    this.replaceChildren(...newFragment.childNodes);
+    return this;
+  }
+}
+
+/**
+ * Creates a fragment in the dom that can easily be replaced by running the
+ * render method of the returned renderable fragment reference.
+ * @param  {string}   id
+ * @param  {Function} template
+ * @param  {Object}   props
+ * @return {Renderable}
+ */
+export function renderable(id, template, props) {
+  return new Renderable(id, template, props);
```

Comment:
this roughly translates to the React.createElement concept.
see https://reactjs.org/docs/jsx-in-depth.html

the `new Renderable` class is there for a simple `instanceof` check later but it could as well return a plain object. The important thing here is that it holds a reference to the template (component) function to re-run later on.